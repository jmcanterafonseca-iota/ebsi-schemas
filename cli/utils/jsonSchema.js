import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import crypto from "node:crypto";
import $RefParser from "@apidevtools/json-schema-ref-parser";
import canonicalize from "canonicalize";
import { globbySync } from "globby";
import inquirer from "inquirer";
// eslint-disable-next-line import/extensions
import { base16 } from "multiformats/bases/base16";

export function removeAnnotations(obj) {
  /**
   * Lists of annotations keywords:
   * - https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.9
   * - https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.9
   * - https://json-schema.org/draft-07/json-schema-validation.html#rfc.section.10
   */
  const keysToRemove = [
    "title",
    "description",
    "default",
    "deprecated",
    "readOnly",
    "writeOnly",
    "examples",
  ];

  return JSON.parse(
    JSON.stringify(obj, (key, val) =>
      keysToRemove.includes(key) ? undefined : val
    )
  );
}

export function replaceRefs(obj, tsr, schemaPath, schemaIds) {
  const currentDir = dirname(fileURLToPath(import.meta.url));
  const rootDir = resolve(currentDir, "../..");

  const dependencies = [];
  const updatedSchema = JSON.parse(
    JSON.stringify(obj, (key, val) => {
      if (key !== "$ref" || typeof val !== "string" || !val.startsWith("../")) {
        return val;
      }

      // Relative path (from schemaPath to absolute path)
      const referencedPath = resolve(schemaPath, "..", val);
      const match = schemaIds.find(({ schema }) => {
        if (resolve(rootDir, schema) === referencedPath) {
          // Add to dependencies array
          if (!dependencies.includes(schema)) {
            dependencies.push(schema);
          }

          return true;
        }
        return false;
      });

      if (!match || !match.id) {
        throw new Error(`Unable to find ${referencedPath}`);
      }

      return `${tsr}/schemas/${match.id}`;
    })
  );

  return {
    updatedSchema,
    dependencies,
  };
}

export async function computeId(schemaPath) {
  const fullPath = resolve(process.cwd(), schemaPath);

  // 1. Bundle schema
  const bundledSchema = await $RefParser.bundle(fullPath);

  // 2. Remove annotations
  const sanitizedDocument = removeAnnotations(bundledSchema);

  // 3. Canonicalise
  const canonicalizedDocument = canonicalize(sanitizedDocument);

  // 4. Compute sha256 of the stringified JSON document
  const hash = crypto
    .createHash("sha256")
    .update(JSON.stringify(canonicalizedDocument), "utf-8")
    .digest();

  // 5. Return multibase base58 (BTC) encoded hash
  // return base58btc.encode(hash);

  // Temporary: until we update TSR API
  return `0x${base16.baseEncode(hash)}`;
}

const currentDir = dirname(fileURLToPath(import.meta.url));
const rootDir = resolve(currentDir, "../..");

export async function getSchemaIds(paths, options) {
  const SCHEMAS_PATTERNS = ["schemas/**/schema.json"];

  let schemas = [];
  if (Array.isArray(paths) && paths.length > 0) {
    schemas = globbySync(paths, { cwd: rootDir });
  } else if (options.all) {
    // Find all schemas
    schemas = globbySync(SCHEMAS_PATTERNS, { cwd: rootDir });
  } else {
    const allSchemas = globbySync(SCHEMAS_PATTERNS, {
      cwd: rootDir,
    });
    const answers = await inquirer.prompt([
      {
        type: "checkbox",
        message: "Which JSON Schema(s) do you want to compute?",
        name: "schemas",
        choices: allSchemas.map((schema) => ({
          name: schema,
        })),
        pageSize: 12,
        validate(answer) {
          if (answer.length < 1) {
            return "You must choose at least one JSON schema.";
          }

          return true;
        },
      },
    ]);
    schemas = answers.schemas;
  }

  if (!(schemas.length > 0)) {
    console.error("No JSON Schema found!");
    process.exit(1);
  }

  return Promise.all(
    schemas.map(async (schema) => {
      const id = await computeId(schema);
      return { id, schema };
    })
  );
}
