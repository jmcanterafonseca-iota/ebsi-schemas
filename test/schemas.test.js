import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import fs from "fs";
import Ajv2020 from "ajv/dist/2020.js";
import Ajv2019 from "ajv/dist/2019.js";
import AjvDraft07 from "ajv";
import addFormats from "ajv-formats";
import $RefParser from "@apidevtools/json-schema-ref-parser";
import { globbySync } from "globby";
// eslint-disable-next-line import/extensions
import { base58btc } from "multiformats/bases/base58";
import { computeId } from "../cli/utils/jsonSchema.js";

// Ignore list: these schemas are not valid ("strict": true, https://ajv.js.org/options.html#strict)
const ignoreList = [
  "schemas/ebsi-muti-uni-pilot/education-verifiable-accreditation-records/2021-12/schema.json",
  "schemas/ebsi-muti-uni-pilot/verifiable-attestation-organisational-id/2021-12/schema.json",
  "schemas/ebsi-muti-uni-pilot/verifiable-diploma/2021-11/schema.json",
  "schemas/ebsi-vid/legal-entity/2021-12/schema.json",
  "schemas/ebsi-vid/natural-person/2021-12/schema.json",
  "schemas/ebsi-vid/verifiable-authorisation/2021-11/schema.json",
];

// Find all schemas (except the broken ones)
const currentDir = dirname(fileURLToPath(import.meta.url));
const rootDir = resolve(currentDir, "..");
const schemas = globbySync(["schemas/**/schema.json"], {
  cwd: rootDir,
  ignore: ignoreList,
});

describe.each(schemas)("Schema %s", (schemaFile) => {
  test("should compile without errors (using Ajv)", async () => {
    // Bundle schema
    const bundledSchema = await $RefParser.bundle(resolve(rootDir, schemaFile));

    // Configure Ajv
    let ajv;
    switch (bundledSchema.$schema) {
      // https://github.com/json-schema-org/json-schema-spec/blob/2020-12/schema.json
      case "https://json-schema.org/draft/2020-12/schema": {
        ajv = new Ajv2020({ allErrors: true, strict: true });
        break;
      }
      // https://github.com/json-schema-org/json-schema-spec/blob/2019-09/schema.json
      case "https://json-schema.org/draft/2019-09/schema": {
        ajv = new Ajv2019({ allErrors: true, strict: true });
        break;
      }
      // https://github.com/json-schema-org/json-schema-spec/blob/draft-07/schema.json
      case "http://json-schema.org/draft-07/schema#": {
        ajv = new AjvDraft07({ allErrors: true, strict: true });
        break;
      }
      default: {
        throw new Error(`Unknown version "${bundledSchema.$schema}"`);
      }
    }
    addFormats(ajv);

    // Should compile without errors
    expect(ajv.compile(bundledSchema)).not.toThrow();
  });
});

test("should match the snapshot", async () => {
  const summary = await Promise.all(
    schemas.map(async (schemaFile) => {
      const idBase16 = await computeId(schemaFile);
      const idBase58 = base58btc.encode(
        Buffer.from(idBase16.replace("0x", ""), "hex")
      );
      const dataString = fs.readFileSync(schemaFile, "utf8");
      const data = JSON.parse(dataString);

      return {
        title: data.title,
        description: data.description,
        idBase16,
        idBase58,
        file: schemaFile,
      };
    })
  );
  expect(summary).toMatchSnapshot();
});
