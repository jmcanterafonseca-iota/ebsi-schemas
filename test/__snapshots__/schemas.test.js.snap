// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema schemas/ebsi-accreditation/2022-01/schema.json should not contain breaking changes 1`] = `"0x7d8700a8eb8c11a7200f9c39ce18b1712ca5d065acf825a2e3a1098f49bd6d6c"`;

exports[`Schema schemas/ebsi-accreditation/2022-02/schema.json should not contain breaking changes 1`] = `"0xb9d2c90ac8041379fd5d682faf4350e74453f89b064f7702d5c9a71e37d7efe8"`;

exports[`Schema schemas/ebsi-accreditation/2022-03/schema.json should not contain breaking changes 1`] = `"0x9c26d19032deb5d72f4ac56f3b88c4025c6be8123c244fe57a7d6e5343b71b65"`;

exports[`Schema schemas/ebsi-accreditation/2022-05/schema.json should not contain breaking changes 1`] = `"0x3b4650f9b09d5ad065c72cec004b8a0896d55e4a25a412cab01ecd6580732803"`;

exports[`Schema schemas/ebsi-accredited-attestation/2022-05/schema.json should not contain breaking changes 1`] = `"0x2dad717a5a0fc9edaf518aeae7f4068fc59487d96fdf576527165b5216303b82"`;

exports[`Schema schemas/ebsi-attestation/2021-11/schema.json should not contain breaking changes 1`] = `"0x7f6c312ff3f7da74376304ac580abc6425e8dfa6009933c8360883049e4e0980"`;

exports[`Schema schemas/ebsi-attestation/2022-02/schema.json should not contain breaking changes 1`] = `"0x28d76954924d1c4747a4f1f9e3e9edc9ca965efbf8ff20e4339c2bf2323a5773"`;

exports[`Schema schemas/ebsi-attestation/2022-05/schema.json should not contain breaking changes 1`] = `"0x8eac64f2689fc98a88800775765102fea5a80939a3caf92e88ee06d364a569b8"`;

exports[`Schema schemas/ebsi-essp/pda1/2022-03/2019-09/schema.json should not contain breaking changes 1`] = `"0x0ce34cab347f891b33b0a7a3e300f67840f5f1eb3a0c76dd2db2991815807fe1"`;

exports[`Schema schemas/ebsi-essp/pda1/2022-03/draft-07/schema.json should not contain breaking changes 1`] = `"0x50490ec4e2b3053f707cc121c26217cbec363f3b9e909175d073d04c73890f9a"`;

exports[`Schema schemas/ebsi-muti-uni-pilot/education-verifiable-accreditation-records/2022-02/schema.json should not contain breaking changes 1`] = `"0x6b304591b86b7e5e4c8a713ddfb84ee5810a56334827adda4e741f80b9aa49ed"`;

exports[`Schema schemas/ebsi-muti-uni-pilot/education-verifiable-accreditation-records/2022-03/schema.json should not contain breaking changes 1`] = `"0x13d597f8495e6b6e3d0c072218756a1bcc3ea50ebeb3ab4c3944bd400e0c3c6a"`;

exports[`Schema schemas/ebsi-muti-uni-pilot/verifiable-attestation-individual-id/2021-12/schema.json should not contain breaking changes 1`] = `"0x5f548a097ba97c3f89d5b4dbc826b3221206ddec19d61cc84b3a756207592985"`;

exports[`Schema schemas/ebsi-muti-uni-pilot/verifiable-attestation-individual-id/2022-02/schema.json should not contain breaking changes 1`] = `"0xb77f8516a965631b4f197ad54c65a9e2f9936ebfb76bae4906d33744dbcc60ba"`;

exports[`Schema schemas/ebsi-muti-uni-pilot/verifiable-attestation-organisational-id/2022-03/schema.json should not contain breaking changes 1`] = `"0x530809afc39dafa06a428c0e166b4678dc8300835edfa7f0cc03c3ac3a61b1e5"`;

exports[`Schema schemas/ebsi-muti-uni-pilot/verifiable-diploma/2022-03/schema.json should not contain breaking changes 1`] = `"0x6b304591b86b7e5e4c8a713ddfb84ee5810a56334827adda4e741f80b9aa49ed"`;

exports[`Schema schemas/ebsi-presentation/2021-11/schema.json should not contain breaking changes 1`] = `"0xff839bbe9c571e93779c0aade78d93b824760dc13ce4b5ea4d9cb0489f99fead"`;

exports[`Schema schemas/ebsi-presentation/2022-02/schema.json should not contain breaking changes 1`] = `"0x7e53dd8c85ffdbf0fff15599e887d6221ab02982d24d857aae1b3d2cc294f048"`;

exports[`Schema schemas/ebsi-vid/legal-entity/2022-02/schema.json should not contain breaking changes 1`] = `"0x3c1a04b86d9840ed2017847932a049d312282ec5d96f4ef86c91ecd6aad1f345"`;

exports[`Schema schemas/ebsi-vid/natural-person/2022-02/schema.json should not contain breaking changes 1`] = `"0xad457662a535791e888994e97d7b5e0cdd09fbae2c8900039d2ee2d9a64969b1"`;

exports[`Schema schemas/ebsi-vid/verifiable-authorisation/2022-02/schema.json should not contain breaking changes 1`] = `"0x9e096019374c3afcb1ae7c4b97afc1aa73bd325afa9557b8cba23ccb11a79895"`;
